From 4709a4e1b21ecc5eb2c01db365ce885f27e60d96 Mon Sep 17 00:00:00 2001
From: Jiangyi Chen <chenjiangyi@uniontech.com>
Date: Thu, 23 Jul 2026 11:40:14 +0800
Subject: [PATCH 236/268] winewayland.drv: Avoid ABBA deadlocks between
 win_data_mutex and user_mutex.

Make the lock acquisition order consistent: always acquire the user_mutex lock
on the user32 side first, and then the win_data_mutex on the wayland side.
Otherwise, when the CCTV Video main program starts, there is a nearly 40%
probability of a race deadlock between set_client_surface and update_surface_region
(which indirectly calls wayland_window_surface_flush).

Signed-off-by: Jiangyi Chen <chenjiangyi@uniontech.com>
---
 dlls/winewayland.drv/wayland_surface.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index a2ed320a886..9c2f7512a22 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -1430,12 +1430,13 @@ static void wayland_client_surface_update(struct client_surface *client)
     struct wayland_client_surface *surface = impl_from_client_surface(client);
     HWND hwnd = client->hwnd, toplevel = NtUserGetAncestor(hwnd, GA_ROOT);
     struct wayland_win_data *data;
+    BOOL visible = FALSE;
 
     TRACE("%s\n", debugstr_client_surface(client));
-
+    if (toplevel) visible = is_client_visible(hwnd);
     if (!(data = wayland_win_data_get(hwnd))) return;
 
-    if (toplevel && is_client_visible(hwnd))
+    if (toplevel && visible)
         wayland_client_surface_attach(surface, toplevel);
     else
         wayland_client_surface_attach(surface, NULL);
@@ -1561,10 +1562,12 @@ void set_client_surface(HWND hwnd, struct wayland_client_surface *new_client)
     HWND toplevel = NtUserGetAncestor(hwnd, GA_ROOT);
     struct wayland_client_surface *old_client;
     struct wayland_win_data *data;
+    BOOL visible = FALSE;
 
     /* ownership is shared with the callers, the last caller to release
      * its reference will also destroy it and clear our pointer. */
 
+    if (toplevel) visible = is_client_visible(hwnd);
     if (!(data = wayland_win_data_get(hwnd))) return;
 
     if (new_client != data->client_surface)
@@ -1574,7 +1577,7 @@ void set_client_surface(HWND hwnd, struct wayland_client_surface *new_client)
 
         if ((data->client_surface = new_client))
         {
-            if (toplevel && is_client_visible(hwnd))
+            if (toplevel && visible)
                 wayland_client_surface_attach(new_client, toplevel);
             else
                 wayland_client_surface_attach(new_client, NULL);
-- 
2.55.0

